File: 美团8.16笔试/t3.cpp
namespace leetcode {
template <typename T> using vector = std::vector<T>;
template <typename T> using function = std::function<T>;

using PII = std::pair<int, int>;

constexpr int N = 2e5 + 2;

int n;
std::array<int, N> a{}, b{}, less_or_equal_right{};

inline int lowbit(int x) { return x & (-x); }
struct TreeArr {
  int upper;
  std::vector<int> arr;
  TreeArr(int _upper) : upper(_upper) { arr.resize(upper + 1); }

  int query(int pos) {
    int res = 0;
    while (pos) {
      res += arr[pos];
      pos -= lowbit(pos);
    }
    return res;
  }

  void update(int pos, int x) {
    while (pos <= upper) {
      arr[pos] += x;
      pos += lowbit(pos);
    }
  }
};

class Solution {
public:
  void solve() {
    freopen("input.txt", "r", stdin);
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);
    std::cout.tie(nullptr);
    std::cin >> n;
    for (int i = 1; i <= n; i++) {
      std::cin >> a[i];
    }
    // 离散化
    std::vector<int> tmp(a.begin(), a.end());
    std::sort(tmp.begin(), tmp.end());
    tmp.erase(std::unique(tmp.begin(), tmp.end()), tmp.end());

    for (int i = 1; i <= n; i++) {
      b[i] = (std::lower_bound(tmp.begin(), tmp.end(), a[i]) - tmp.begin()) + 1;
    }

    long long ans = 0;
    TreeArr backward_tree = TreeArr(tmp.size() + 1);
    for (int i = n; i >= 1; i--) {
      int cnt = backward_tree.query(b[i] - 1);
      less_or_equal_right[i] = backward_tree.query(b[i]);
      ans += static_cast<long long>(cnt) * (cnt - 1) / 2;
      backward_tree.update(b[i], 1);
    }
    TreeArr forward_tree = TreeArr(tmp.size() + 1);
    for (int i = 1; i <= n; i++) {
      int cnt = i - 1 - forward_tree.query(b[i]);
      ans -= static_cast<long long>(cnt) * less_or_equal_right[i];
      forward_tree.update(b[i], 1);
    }
    std::cout << ans << std::endl;
  }
};
} // namespace leetcode

File: 美团8.16笔试/t3.py
import sys
import bisect

def low_bit(x):
    return x & (-x)

class TreeArr:
    def __init__(self, upper):
        self.arr = [0] * (upper + 1)
        self.upper = upper

    def add(self, pos, val):
        now = pos
        while (now <= self.upper):
            self.arr[now] += val
            now += low_bit(now)

    def sum(self, pos):
        now = pos
        res = 0
        while now:
            res += self.arr[now]
            now -= low_bit(now)
        return res

def main():
    input = sys.stdin.read().strip().split()
    it = iter(input)
    n = int(next(it))
    a = [0] * (n + 1)
    for i in range(1, n + 1):
        a[i] = int(next(it))
    # 离散化
    vals = sorted(set(a))
    _max = len(vals)
    func = lambda x: bisect.bisect_left(vals, x) + 1 
    arr = [func(v) for v in a] # 注意这里是a，不是vals，把a里面大的值域映射到小的值域

    ans = 0
    # 我右边，小于等我的个数，用来计算a_j >= a_k
    less_or_equal_right = [0] * (n + 1)
    # 我左边，大于我的个数，用来计算a_i>a_j
    # 因为树状数组只能向下统计，所以这个greater_left需要通过一个减法间接得到
    # 不需要真的创建出来，可以在第二个循环逐步使用
    # greater_left = [0] * (n + 1)

    # 从后往前遍历
    backward_ta = TreeArr(_max)
    for i in range(n, 0, -1):
        cnt = backward_ta.sum(arr[i] - 1)
        less_or_equal_right[i] = backward_ta.sum(arr[i])
        ans += cnt * (cnt - 1) // 2 # 即组合数C(n, 2)
        backward_ta.add(arr[i], 1)
    
    # 然后从前往后
    forward_ta = TreeArr(_max)
    for i in range(1, n + 1, 1):
        cnt = i - 1 - forward_ta.sum(arr[i])
        ans -= cnt * less_or_equal_right[i]
        forward_ta.add(arr[i], 1)

    print(ans)


if __name__ == "__main__":
    main()
